Discussion fixes from solutions

e-kolpakov opened the pull request:
**Description:** Solutions fork contains some fixes/improvements  for Discussions JS code - this PR aims to pull them into upstream edx-platform

**JIRA:** [YONK-328](https://openedx.atlassian.net/browse/YONK-328)

**Sandboxes:** [LMS](http://pr12914.sandbox.opencraft.hosting/), [Studio](http://studio-pr12914.sandbox.opencraft.hosting/)

**Testing instructions:**

Short: No changes in functionality are added - running test suite and smoke-testing everything as single user should be enough.

Detailed:
1. Create&publish a course
2. Log in to LMS, go to course
3. Go to "Discussions" tab
4. Create two-three new posts of different types (question/discussion)
5. Check navigation works - jump between threads. Refresh the page and make sure the same thread is displayed after refresh
6. (optionally, log in as a different user) Add some comments and responses.
7. Try deleting comments/responses/threads
8. Log in as different user - try upvoting threads
9. As thread/response/comment owner - try editing contents

Reviewers:

* [x] @bradenmacdonald 
* [x] @robrap 
* [x] @bjacobel

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=12914

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -18,9 +18,20 @@
 
+            var allThreadsRoute = "",
+                singleThreadRoute = getSingleThreadRoute(":forum_name", ":thread_id"), // :forum_name/threads/:thread_id
+                routes = [];
```
should be `routes = {};` if it's going to be an object with string keys.

Also, just FYI, once edX switches to Babel / ES6 you can write this as:
```
DiscussionRouter.prototype.routes = {
    [allThreadsRoute]: "allThreads",
    [singleThreadRoute]: "showThread",
};
```

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -127,17 +135,14 @@
-                    trigger: true
-                });
+                targetThreadRoute = getSingleThreadRoute(thread.get("commentable_id"), thread_id);
+                this.navigate(targetThreadRoute, { trigger: true});
```
Do you know for sure that the `return` value was never being used?

e-kolpakov replied:
@bradenmacdonald yes:
* in [Backbone docs](http://backbonejs.org/#Router-navigate) navigate return value is never used
* it's just coffeescript always returning the result of last line of code

bjacobel replied:
uneven spacing inside `{ trigger: true}`

bjacobel replied:
(FYI our new JS styleguide says not to add extra space inside single-line objects, e.g., `{trigger: true}` - but we're not enforcing the new style in edx-platform yet, so for this PR either spaces to either side or no spaces is fine, the eslint autofixer will switch them all to no-spaces later)

bradenmacdonald commented on common/static/common/js/discussion/main.js:
```diff
@@ -33,10 +33,12 @@
+                if (!Backbone.History.started) {
+                    Backbone.history.start({pushState: true, root: "/courses/" + $$course_id + "/discussion/forum/"});
+                }
+                else {
```
Nit: `} else {`

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -102,11 +108,43 @@
+                        self.thread = thread;
+                        self.renderThreadView();
+                    };
+                    this.retrieveSingleThread(forum_name, thread_id, callback);
```
Tiny nit (you can ignore this whole comment if you want): I find this easier to follow when you ditch the intermediate variable and move the code into the order in which it gets called:

```
this.retrieveSingleThread(forum_name, thread_id, function(thread) {
     self.thread = thread;
     self.renderThreadView();
});
```

Or, if you wanted, instead of accepting a `callback` argument, you could make `retrieveSingleThread` so it creates returns `$.Deferred()`, and calls `deferred.resolve(new Thread(data['content']))` within the ajax `success` handler. Then you can do `this.retrieveSingleThread(forum_name, thread_id).done(function(thread) { ... });` which is even more clear.

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -102,11 +108,43 @@
+                    };
+                    this.retrieveSingleThread(forum_name, thread_id, callback);
+                }
+                else {
```
Nit: `} else {`

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -102,11 +108,43 @@
+                self = this;
                 this.thread = this.discussion.get(thread_id);
+
+                if (!this.thread){
```
Can you add a comment explaining what circumstances could cause `this.thread` to be falsy?

e-kolpakov replied:
@bradenmacdonald Unfortunately that's one feature @Kelketek implemented, and there aren't any references to PR on the [commit](https://github.com/edx-solutions/edx-platform/commit/e28dde32169b388235030dba945366f4c13e565f) for some reason. So the motivation is unknown (yet).

@Kelketek is there any chance you still remember what was the reason for this?

Kelketek replied:
@e-kolpakov I remember this would happen when users got an email pointing to a specific thread. If the thread wasn't on the first page of results when loading the discussion, it would break. We had this happen a few times-- we'd apply the patches, rebase, this fix would go away, the problem would happen again, and we'd put it back.

e-kolpakov replied:
@Kelketek cool, thank you - looks like we should finally add a test for it.

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -102,11 +108,43 @@
+                    success: function(data) { callback(new Thread(data['content'])); },
+                    error: function(xhr) {
+                        var errorMessage = (xhr.status === 404) ?
+                            gettext("The thread you selected has been deleted. Please select another thread.") :
```
Is deletion the only likely cause of a 404?

e-kolpakov replied:
@bradenmacdonald [yes](https://github.com/edx/edx-platform/blob/4fd7f2936aefe9421c704c266d3f35bb3ca7f1a5/lms/djangoapps/django_comment_client/forum/views.py#L323).

Well, in theory, it will also happen if discussions are disabled, but we won't get to this code than in the first place. Another option is changed URL, but that would make discussions totally broken. So, yes, there's only one legitimate condition for 404 - deleted thread

bradenmacdonald commented on common/static/common/js/discussion/views/discussion_thread_view.js:
```diff
@@ -86,13 +86,6 @@
-                    if (collection.get(id)) {
-                        self.model = collection.get(id);
-                    }
-                });
```
What did this code do and why did you delete it?

e-kolpakov replied:
@bradenmacdonald same comment as above

e-kolpakov replied:
@Kelketek one more question: this it was deleted in the same comment that added "retrieveSingleThread" - any chance they are related?

Kelketek replied:
@e-kolpakov I'm afraid I don't remember the state of things well enough to tell you that. It's likely-- I know since we had to re-add this a few times I probably tried to isolate as small a change as I could to get the desired effect. Don't know how much has changed since then, either.

bradenmacdonald replied:
@e-kolpakov  I'd rather not include code changes we don't understand, so please see if there is any difference with or without this change.

Thanks for stopping by, @Kelketek  :)

bradenmacdonald commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -137,7 +137,8 @@
                         'message': searchAlert.attributes.message,
-                        'cid': searchAlert.cid
+                        'cid': searchAlert.cid,
+                        'css_class': searchAlert.attributes.css_class
```
Nit: you can add a trailing comma for nicer diffs in the future.

bradenmacdonald commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -153,11 +154,12 @@
             };
 
-            DiscussionThreadListView.prototype.addSearchAlert = function(message) {
+            DiscussionThreadListView.prototype.addSearchAlert = function(message, css_class) {
```
Can you add a comment / [JSDoc](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/javascript_guidelines.html#documentation) explaining the purpose of css_class? What's an example use case?

bradenmacdonald commented on common/static/common/js/spec/discussion/view/new_post_view_spec.js:
```diff
@@ -94,7 +94,8 @@
                 DiscussionSpecHelper.makeAjaxSpy(function(params) {
-                    return expect(params.data.group_id).toEqual(expectedGroupId);
+                    expect(params.data.group_id).toEqual(expectedGroupId);
+                    self.view.$(".forum-new-post-form").removeAttr("disabled");
```
This seems like a weird side effect for a method called `makeAjaxSpy`. Can you put this somewhere else?

bradenmacdonald commented on common/static/common/js/spec/discussion/view/response_comment_view_spec.js:
```diff
@@ -164,10 +166,17 @@
+                    // element.after(...);
+                    // element is event.target in this case. This causes a JS exception, so we override the target
+                    this.event.target = $("body");
+                    $(this.event.target).prop("disabled", false);
```
This is a bit messy. It would be nicer if you modify `makeEventSpy` so that after creating the `event` mock, it assigns `event.target.after = createSpy('after')` before returning `event`.

e-kolpakov replied:
@bradenmacdonald unfortunately, no - it is not passed straight to utils.showLoadingIndicator - instead it is wrapped into jQuery object a [couple](https://github.com/edx/edx-platform/blob/4fd7f2936aefe9421c704c266d3f35bb3ca7f1a5/common/static/common/js/discussion/views/response_comment_view.js#L142) of [times](https://github.com/edx/edx-platform/blob/4fd7f2936aefe9421c704c266d3f35bb3ca7f1a5/common/static/common/js/discussion/utils.js#L205), so it ends up being a document again (not a spy).

bradenmacdonald commented on lms/templates/discussion/index.html:
```diff
@@ -22,6 +22,13 @@
+           DiscussionApp.start(elem, true);
+       });
+   });
+ </script>
```
What is the purpose of this change? To allow loading this by injecting its HTML into a page after main.js has loaded and the DOM ready event has already fired?

e-kolpakov replied:
@bradenmacdonald mostly to reduce code drift - but we need it in solutions so that the 3rd party tool that should not be named could work; can't recall the details, but probably jquery.xblock had some trouble with this.

bjacobel replied:
I'm unclear on why this is needed (and what the "the 3rd party tool that should not be named" is), but we should avoid script tags in HTML if at all possible. It's not testable (or lintable) and takes us further from the goal of getting all of Discussions to be loaded with RequireJS.

The better solution is to create a factory and load it with the mako `require_module` function. [See Andy's `DiscussionBoardFactory` here](https://github.com/edx/edx-platform/pull/12772/files#diff-b9a60134e0d9d993486faace7ec1513fR31). Again, not really clear what exactly this does so I can't give more specific direction at the moment, but I'll see what I can do if you fill me in a little more.

e-kolpakov replied:
@bjacobel "the 3rd party tool that should not be named" is a 3rd party tool some solutions client uses instead of LMS. And most of the fixes/changes I'm "forward-porting" in this PR are made to fix bugs caused or exposed by the use of that 3rd party tool.

I've been lectured a couple of times in the past to not mention clients of edX in public places like github and Slack, so I'm not naming it. If you're curious, ask Steve Magoun - he knows for sure who I am talking about.

Regarding this partiicular change - the only actual reason to bring it here is to reduce code drift. But it is relatively simple, so I believe such code drift is affordable. It looks like the best course of actions would be to remove this commit from this PR - I'll do that now and if you have any better ideas - let me know.

bradenmacdonald commented on common/static/common/js/discussion/utils.js:
```diff
@@ -194,18 +194,6 @@
             params.url = URI(params.url).addSearch({
                 ajax: 1
             });
-            params.beforeSend = function() {
```
Is there any test that verifies that a global beforeSend is respected after this change? If not, you should add one.

e-kolpakov replied:
@bradenmacdonald added in https://github.com/edx/edx-platform/pull/12914/commits/24b27fab925c2f93512aaa6d92606e0a5a494ace

bradenmacdonald commented:
Can you explain the change "goHome no longer triggers thread removal" please?

e-kolpakov commented:
@bradenmacdonald goHome [used](https://github.com/edx/edx-platform/pull/12914/commits/9768bfbe55ec6e469af6271438239c8064d440f9#diff-63fbfdb65388a776cb73ce60b8908d06L427) side-effect of thread:removed event [to jump to "home page"](https://github.com/open-craft/edx-platform/blob/9768bfbe55ec6e469af6271438239c8064d440f9/common/static/common/js/discussion/discussion_router.js#L64). This was quite counter-intuitive, so instead of triggering "thread:removed" it now just does not do anything.

It is confusing and also does not actually do anything, as thread:removed was already triggered in threadRemoved; maybe it was stale code.

In the same commit there's a change to keep track of current displayed thread and only delete thread if it is selected, but after taking a closer look I don't think it is a good idea (and can't recall why I've done so in the first place), so I'll remove it.

e-kolpakov commented:
@bradenmacdonald please take another look

bradenmacdonald commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -151,11 +153,18 @@
+            /**
+             * Creates search alert model and adds it to collection
+             * @param message - alert message
+             * @param css_class - css class for a message
```
Nit: "css class for a message" isn't a very helpful explanation :-p . That says what but not why.

bradenmacdonald commented:
@e-kolpakov LGTM and code seems to be working. 👍 

This is going to be a bit hard for subsequent reviewers to follow each related change though. Can you please squash your fixups into the relevant commits, or add a list of the changes or something? Then go ahead and ping for review.

e-kolpakov commented:
jenkins run python please

e-kolpakov commented:
@bradenmacdonald done.

@robrap please review.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -263,9 +278,9 @@
+                var makeErrorElem, response, _i, _len, _ref, _results, $errorItem;
                 makeErrorElem = function(message) {
-                    return $("<li>").addClass("post-error").html(message);
+                    return $("<li>").addClass("post-error").html(edx.HtmlUtils.ensureHtml(message).toString());
```
I think you could avoid the `'toString()` by using (untested):
```
return edx.HtmlUtils.setHtml(
    $("<li>").addClass("post-error"),
    message
);
```
I think we'd prefer using HtmlUtils calls that have `ensureHtml` built in where they exist (like `setHtml`), and save the above style (which is accurate) for JQuery functions that aren't supported directly in HtmlUtils.

robrap commented:
I'm reviewing the safe template work.  
@bjacobel: It would be great if you could help on the general JavaScript.  Thanks.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -151,61 +153,59 @@
+                            edx.HtmlUtils.HTML("<button class='dismiss'>"),
+                                gettext("OK"),
+                            edx.HtmlUtils.HTML("</button>"),
+                        edx.HtmlUtils.HTML("</div>")
```
This is accurate.  Ideally, larger HTML blocks would move to Underscore.js files.  I like that this was tabbed so that it is mostly readable when concentrating on the HTML itself.  Another style that has been used can be seen here:
https://github.com/edx/edx-platform/blob/7d8f2e95785ac94f4db4253d68b67def46e948f8/common/lib/xmodule/xmodule/js/src/video/09_video_caption.js#L84-L118

@bjacobel @andy-armstrong: When you get a chance, maybe you can weigh in on the style we'd like to lean towards for this type of code.  We may also want to update the documentation with our preferences.

Thanks.

bjacobel replied:
I'm against HTML living inside Javascript generally; HTML should go in an Underscore template and be loaded into Javascript with RequireJSText, if that's an option (it may not be here, it looks like this code isn't RequireJS-ified yet).

That said, of the two patterns for writing HTML inside JS, I prefer the one @robrap linked to instead of the one that's shown here - if there is HTML in the JS, it should at least be readable as a multiline block of plain HTML, and not interspersed with JS on each line. I don't see the pattern above as much of a win for readability - it repeats itself frequently and removes the indentation of the HTML, which is key to understanding its structure and purpose. Perhaps this change was done for a different reason though?

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -353,31 +370,38 @@
             processedText = "";
-            $div = $("<div>").html(text);
+            // safe-lint: disable=javascript-jquery-html
+            $div = $("<div>").html(safeHtml);
```
You should use `ensureHtml()` if safeHtml is an HTML Snippet that was made safe by another `HtmlUtils` call.  It will be a no-op, but ensure that `HtmlUtils` will have been called.  If it isn't, can you turn it into real `safeHtml` before passing it to this function?

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -386,26 +410,28 @@
-            $div = $("<div>").html(text);
+            safeHtml = safeHtml.replace(new RegExp(ESCAPED_BACKSLASH, 'g'), '\\\\\\\\');
+            // safe-lint: disable=javascript-jquery-html
+            $div = $("<div>").html(safeHtml);
```
See other comment about ensuring this is safe:
https://github.com/edx/edx-platform/pull/12914/files#r69576197

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -386,26 +410,28 @@
             cnt = 0;
             $div.find("code").each(function(index, code) {
+                // safe-lint: disable=javascript-jquery-html
                 $(code).html(processor(codeArchive[cnt], 'code'));
```
I'd prefer seeing any of the many options other than disabling the rule.  
1. Could `processor(codeArchive[cnt], 'code')` return an HTML Snippet (using HtmlUtils)?  This is typically the best option if possible.
2. If this is not possible, I prefer `HtmlUtils.HTML(processor(codeArchive[cnt], 'code'))` to declare this as HTML rather than disabling the linter.  This basically does the same thing, which is why option #1 is better, but I prefer seeing the solution in code than in a comment where possible.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -421,15 +447,16 @@
             text = text.replace(/^\&gt\;/gm, ">");
             converter = Markdown.getMathCompatibleConverter();
-            text = this.unescapeHighlightTag(this.stripLatexHighlight(converter.makeHtml(text)));
+            var safeHtml = converter.makeHtml(text);
```
For example, if you know that `converter.makeHtml(text)` returns safe html, then you could use:
```
var safeHtml = HtmlUtils.HTML(converter.makeHtml(text));
```
In this way, safeHtml would be turned into an HTML Snippet in the context in which you can see that is safe to do, and then it can be used with other HtmlUtils functions from this point out (or using toString() where and when you must).

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -446,17 +473,23 @@
         DiscussionUtil.abbreviateHTML = function(html, minLength) {
             var $result, imagesToReplace, truncated_text;
-            truncated_text = jQuery.truncate(html, {
+            truncated_text = edx.HtmlUtils.HTML(jQuery.truncate(html, {
```
Is `html` really HTML?  If so, could it be set to an HTML Snippet using `HtmlUtils` before being passed to this function?  To use `JQuery.truncate()`, you might then need to leave this line as-is and add `html.toString()`, but at least you would have some context for declaring this as HTML().  

If html is really plain text, you could rename it and just drop this call to `HtmlUtils.HTML()`.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -446,17 +473,23 @@
             if (imagesToReplace.length > 0) {
-                $result.append("<p><em>Some images in this post have been omitted</em></p>");
+                $result.append(
+                    edx.HtmlUtils.HTML("<p><em>Some images in this post have been omitted</em></p>").toString()
```
This needs `HtmlUtils.joinHtml()` after translating text using `gettext()`.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -446,17 +473,23 @@
+                );
             }
-            imagesToReplace.replaceWith("<em>image omitted</em>");
+            imagesToReplace.after(edx.HtmlUtils.HTML("<em>image omitted</em>").toString()).remove();
```
This needs `HtmlUtils.joinHtml()` after translating text using `gettext()`.

robrap commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -291,19 +300,28 @@
+                                edx.HtmlUtils.HTML("<a href='#' class='forum-nav-load-more-link'>"),
+                                    gettext("Load more"),
+                                edx.HtmlUtils.HTML("</a>"),
+                            edx.HtmlUtils.HTML("</li>")
```
Same formatting question as: https://github.com/edx/edx-platform/pull/12914/files#r69571560
Let's see how others feel.

bjacobel replied:
Commented on this pattern at that link.

robrap commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -291,19 +300,28 @@
+                        edx.HtmlUtils.HTML('<span class="sr" role="alert">'),
+                        edx.HtmlUtils.ensureHtml(srText),
+                        edx.HtmlUtils.HTML('</span>'),
+                        edx.HtmlUtils.HTML('</div>')
```
Same formatting question as: https://github.com/edx/edx-platform/pull/12914/files#r69571560
Let's see how others feel.

bjacobel replied:
Commented on this pattern at that link.

robrap commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -505,19 +521,19 @@
-                width = test.width();
-                test.remove();
+                $("body").append($test);
+                $test.html(edx.HtmlUtils.ensureHtml(name).toString());
```
Is name plain text?  If so, I'd use `$test.text(name)`.  Otherwise, this is fine, or you could use `HtmlUtils.setHtml()`.

robrap commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -731,18 +755,25 @@
+                                        '<a class="link-jump" href="{url}">',
+                                        {url: DiscussionUtil.urlFor("user_profile", response.users[0].id)}
+                                    )
+                                ),
```
The url should be escaped, so you want something like (untested):
```
edx.HtmlUtils.joinHtml(
    edx.HtmlUtils.HTML('<a class="link-jump" href="'),
    DiscussionUtil.urlFor("user_profile", response.users[0].id),
     edx.HtmlUtils.HTML('">')
)
```
Or you could use `edx.HtmlUtils.interpolateHtml()` and wrap the first param with `HtmlUtils.HTML()` and interpolate in the url as you did above.

robrap commented on common/static/common/js/discussion/views/discussion_thread_view.js:
```diff
@@ -312,22 +315,22 @@
 
                     responsePagination.append($("<span>")
-                        .addClass("response-display-count").html(_.escape(showingResponsesText)));
+                        .addClass("response-display-count").text(_.escape(showingResponsesText)));
```
`.text()` escapes, so you'll want to remove the call to `_.escape()`.

robrap commented on common/static/common/js/discussion/views/discussion_thread_view.js:
```diff
@@ -312,22 +315,22 @@
-                        loadMoreButton = $("<button>").addClass("load-response-button").html(_.escape(buttonText));
-                        loadMoreButton.click(function() {
-                            return self.loadResponses(responseLimit, loadMoreButton);
+                        $loadMoreButton = $("<button>").addClass("load-response-button").text(_.escape(buttonText));
```
`.text()` escapes, so you'll want to remove the call to `_.escape()`.

robrap commented:
Thanks @e-kolpakov.  As you can see from the comment from @bjacobel, we seem to have multiple projects going in the same area, and unfortunately ours is on a feature branch.  @dianakhuang mentioned she could help with merge conflicts.  Of course we should only deal with safe template work in one place.  Hopefully @dianakhuang and @bjacobel can help coordinate a plan to not duplicate work.

e-kolpakov commented:
jenkins run bokchoy please

e-kolpakov commented:
@dianakhuang @bjacobel @andy-armstrong so, let's coordinate on how we will resolve conflicts between this PR and #12772.

It looks like this PR is almost ready to be merged (at least it is mine impression), so do you think it would be better if we merge this first, than rebase #12772 on top of this?

@bjacobel also, still wait for your review of JS changes.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -291,21 +311,21 @@
                 if (type === 'display') {
                     return s.replace(RE_DISPLAYMATH, function($0, $1) {
-                        return "\\[" + $1 + "\\]";
+                        return edx.HtmlUtils.HTML("\\[" + $1 + "\\]");
```
Seems like the usage of the processor below assumes it is taking a string (that contains HTML) and would return a string (that contains HTML).  I think this code is converting to an HtmlSnippet, and then the concatenation below is just turning that back to a string.

I think I'd remove these 3 calls to `edx.HtmlUtisl.HTML`, unless I am missing something.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -353,31 +373,38 @@
+        /**
+         * Formats math and code chunks
+         * @param safeHtml - post contents in form of safe (escaped and/or stripped) HTML
+         * @param processor - callback to post-process math and code chunks. Should return HtmlUtils.HTML or "subclass"
```
See comment above: https://github.com/edx/edx-platform/pull/12914/files#r69921581

I'm not sure this makes sense to return an HtmlUtils.HTML if it doesn't take an HtmlSnippet.  It just seems like the input and output should be consistent, since its purpose doesn't seem to be to convert plain text into HTML.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -353,31 +373,38 @@
             });
-            text = $div.html();
-            text = text.replace(/\\\$/g, ESCAPED_DOLLAR);
+            tempText = $div.html();
```
Maybe `htmlString` would be a better name?

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -386,51 +413,55 @@
                     });
                 } else {
-                    processedText += text;
+                    processedText += tempText;
```
Same naming question for `processedText`.  Maybe `processedHtmlString`?

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -386,51 +413,55 @@
-                return cnt += 1;
+                edx.HtmlUtils.setHtml(
+                    $(code),
+                    edx.HtmlUtils.ensureHtml(processor(codeArchive[index], 'code'))
```
I see that this would require processor to return an HTML Snippet, or to be wrapped with an HtmlUtil.HTML() call instead.

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -386,51 +413,55 @@
+            tempText = tempText.replace(new RegExp(ESCAPED_DOLLAR, 'g'), '\\$');
+            tempText = tempText.replace(/\\\\\\\\/g, ESCAPED_BACKSLASH);
+            tempText = tempText.replace(/\\begin\{([a-z]*\*?)\}([\s\S]*?)\\end\{\1\}/img, function($0, $1, $2) {
                 return processor(("\\begin{" + $1 + "}") + $2 + ("\\end{" + $1 + "}"));
```
Same issue here.  My main issue with processor is that it seems like it should take and return and HTML Snippet, or take and return a String with HTML, but doesn't seem right to take one and return the other.  I'm just repeating myself now.

robrap commented on common/static/common/js/discussion/views/discussion_content_view.js:
```diff
@@ -277,14 +277,16 @@
+                        button.find(".js-sr-vote-count").empty().text(
+                            edx.StringUtils.interpolate(votesCountMsg, {numVotes: numVotes })
+                        );
+                        votesHtml = edx.StringUtils.interpolate(
```
votesHtml => votesText

robrap commented on common/static/common/js/discussion/views/discussion_thread_profile_view.js:
```diff
@@ -42,7 +42,10 @@
-                this.$el.html(_.template($("#profile-thread-template").html())(params));
+                edx.HtmlUtils.setHtml(
+                    this.$el,
+                    edx.HtmlUtils.HTML(_.template($("#profile-thread-template").html())(params))
```
Can you use `HtmlUtils.template()`?

robrap commented on common/static/common/js/discussion/views/discussion_thread_show_view.js:
```diff
@@ -46,24 +46,27 @@
                     readOnly: $('.discussion-module').data('read-only')
                 }, this.model.attributes);
-                return this.template(context);
+                return edx.HtmlUtils.HTML(this.template(context));
```
Could you use HtmlUtils.template instead?

robrap commented on common/static/common/js/discussion/views/response_comment_show_view.js:
```diff
@@ -38,11 +38,13 @@
-                }, this.model.attributes)));
+                }, this.model.attributes));
+
+                edx.HtmlUtils.setHtml(this.$el, edx.HtmlUtils.HTML(renderedTemplate));
```
Same HtmlUtils.template() question.

robrap commented on common/static/common/js/discussion/views/thread_response_show_view.js:
```diff
@@ -41,11 +41,11 @@
                     readOnly: $('.discussion-module').data('read-only')
                 }, this.model.attributes);
-                return this.template(context);
+                return edx.HtmlUtils.HTML(this.template(context));
```
Same HtmlUtils.template() question.

bjacobel commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -18,9 +18,20 @@
+                singleThreadRoute = getSingleThreadRoute(":forum_name", ":thread_id"), // :forum_name/threads/:thread_id
+                routes = {};
+
+            routes[allThreadsRoute] = "allThreads";
```
Prefer using dot notation to set object properties when the property name allows it (as these do), e.g. `routes.allThreadsRoute`

bjacobel replied:
Oh, I'm dumb, these are variables not property names. Ignore.

bjacobel commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -89,16 +100,44 @@
                     return this.nav.setActiveThread(this.thread.get("id"));
                 } else {
-                    return this.nav.goHome;
+                    return this.nav.goHome();
```
This change seems worrying to me... was this previously broken?

e-kolpakov replied:
@bjacobel Yes and no. Yes - the intent here was to show discussion main page

![image](https://cloud.githubusercontent.com/assets/3330048/16683001/2c6d44b2-4506-11e6-89e9-1b0eb11147b1.png)

However, goHome is a method, and setActiveThread is called from Backbone event handler that  does not expect to get anything returned, so return value is never used.

On the other hand, the problem was masked by some code that is executed later - it triggered `allThreads` method above, which executes `goHome`, effectively masking the problem.

One more thing - looks like this else clause is a "almost" dead code anyway, as it is only executed during initialization (`this.thread` is never unset, even after thread is deleted). and it basically did nothing for years and no one noticed. I'll just remove from here.

bjacobel commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -89,16 +100,44 @@
 
             DiscussionRouter.prototype.showThread = function(forum_name, thread_id) {
+                var self;
+                self = this;
```
May as well do this on one line.

e-kolpakov replied:
@bjacobel good point. To defend myself, this is cherry-picked from fix that was made in coffee, and this is how it coffee gernerates code :)

bjacobel commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -89,16 +100,44 @@
+                    self.renderThreadView();
+                }).fail(function(xhr) {
+                    // otherwise display error message and navigate to all threads view
+                    var errorMessage = (xhr.status === 404) ?
```
Multiline ternaries are difficult to parse (for me at least), suggest just using a conventional `if` here.

bjacobel commented on common/static/common/js/discussion/main.js:
```diff
@@ -2,10 +2,9 @@
 
     if (typeof Backbone !== "undefined" && Backbone !== null) {
-        DiscussionApp = {
+        window.DiscussionApp = {
```
Why is this needed? Would prefer to keep things off of `window` unless totally unavoidable.

e-kolpakov replied:
@bjacobel to give visibility that they are on `window` anyway. Note there are no var declaration for `DiscussionApp`, so they are essentially declared in global namespace. With `window` in front this smell is more explicit. 

And I'm afraid fixing it at this point does not make much sense - proper way to do that would be to use RequireJS, and since it is used in DiscussionXBlock we can't do that yet.

bjacobel commented on common/static/common/js/discussion/utils.js:
```diff
@@ -151,61 +153,59 @@
 
         DiscussionUtil.discussionAlert = function(header, body) {
-            var alertDiv, alertTrigger;
+            var $alertDiv, $alertTrigger;
```
:+1: for making sure jQuery objects get a `$` in front, nice.

bjacobel commented on common/static/common/js/discussion/utils.js:
```diff
@@ -151,61 +153,59 @@
                 self = this;
             $elem = params.$elem;
-            if ($elem && $elem.attr("disabled")) {
+            if ($elem && $elem.prop("disabled")) {
```
:+1:

Just curious, was this broken, or just a cleanup fix? Hopefully jQuery Migrate is doing its job at backwards-supporting stuff like this.

e-kolpakov replied:
@bjacobel both. I recall there was some weird edge case when it didn't work as expected, but what was that exactly I don't remember. It might have something to do with 3rd party LMS one solutions client are using.

bjacobel commented on common/static/common/js/discussion/utils.js:
```diff
@@ -216,9 +216,21 @@
             request = $.ajax(params).always(function() {
                 if ($elem) {
-                    $elem.removeAttr("disabled");
+                    $elem.prop("disabled", false);
```
:thumbsup:

bjacobel commented on common/static/common/js/discussion/utils.js:
```diff
@@ -216,9 +216,21 @@
+                if (params.loadingCallback) {
+                    params.loadingCallback.apply(params.$loading);
+                } else {
+                    self.showLoadingIndicator($(params.$loading), params.takeFocus);
```
Curious why `$(params.$loading)` is needed - is `params.$loading` not already a jquery object?

bjacobel commented on common/static/common/js/discussion/utils.js:
```diff
@@ -241,6 +253,9 @@
             }
             undo = _.pick(model.attributes, _.keys(updates));
             model.set(updates);
+            if (typeof beforeSend === 'function') {
```
Good call doing this check first.

bjacobel commented on common/static/common/js/discussion/utils.js:
```diff
@@ -444,19 +475,45 @@
-                $result.append("<p><em>Some images in this post have been omitted</em></p>");
+                edx.HtmlUtils.append(
+                    $result,
+                    edx.HtmlUtils.joinHtml(
```
A few more instances of the HTML pattern Robert and I commented on above down in this file.

bjacobel commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -236,12 +246,14 @@
-                var self = this;
-                this.timer = 0;
-                this.$el.html(this.template({
+                var self = this, $elem;
```
Nit: new JS style guide (which as I said above hasn't been turned on here yet so it's not totally fair for me to review things against it...) says to [not do initialization and instantiation on the same line](https://github.com/edx/eslint-config-edx#one-var-declaration-per-line). Since you assign a value to `$elem` on the next line, may as well just:

```javascript
var self = this,
    $elem = this.template({
        ...
    });
```

bjacobel commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -258,20 +270,17 @@
-                    rendered.append(content);
+                var $content, thread, i, len;
+                this.$(".forum-nav-thread-list").empty();
+                for (i = 0, len = this.displayedCollection.models.length; i < len; i++) {
```
Thanks for cleaning this up, the Coffeescript generated code here was a little gross.

bjacobel commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -392,43 +411,40 @@
-                        } else {
-                            return $('input.email-setting').removeAttr('checked');
-                        }
+                        $('input.email-setting').prop('checked', response.status);
```
:thumbsup:

bjacobel commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -765,23 +794,17 @@
-                    });
-                }
+                var checkbox, checked, urlName;
+                checkbox = $('input.email-setting');
```
`checkbox` should be `$checkbox`.

bjacobel commented on common/static/common/js/spec/discussion/view/discussion_thread_list_view_spec.js:
```diff
@@ -261,14 +261,23 @@
             };
 
+            getAlertMessagesAndClasses = function () {
+                return $(".search-alert").map(function() {
```
inconsistent spacing between `function` and `()` on these two lines

(As mentioned above, new styleguide/inter will catch stuff like this, the preferred style is no space (`function() {`)  - but just be consistent one way or another and the autofixer will get stuff like this when we turn it on in this repo)

bjacobel commented on lms/templates/discussion/user_profile.html:
```diff
@@ -18,6 +19,13 @@
+           DiscussionProfileApp.start(elem);
+       });
+   });
+</script>
```
See my comments here: https://github.com/edx/edx-platform/pull/12914/files#r69946981

bjacobel commented:
Done with my initial review.

e-kolpakov commented:
jenkins run all

e-kolpakov commented:
@robrap @bjacobel I'm afraid I feel this PR scope spirals out of control due to all this safe template fixes - each new fix/improvement/better-than-previous-fix introduces changes to previously intact files, so I have to lint them too, and so on. As a result, we're loosing track of what actually changed: the first 7 commits (the reason this PR exists) only touched 10 files with ~200 lines changes, but now it is already 23 files and more than 600 lines changed.

And we haven't got to resolving merge conflicts with #12772.

@bjacobel @andy-armstrong @dianakhuang friendly reminder to plan resolving merge conflicts between this PR and #12772.

e-kolpakov commented:
jenkins run python
jenkins run a11y

And please make sure there are no errors in flaky tests

e-kolpakov commented:
jenkins run python

e-kolpakov commented:
jenkins run js

e-kolpakov commented:
jenkins run python

e-kolpakov commented:
jenkins run js

robrap commented:
@e-kolpakov: Thank you for all the help with the safe templates fixes.  As a general rule, testing safe templates changes along with other changes saves testing time, but this isn't always true if the scope gets way out of sync.

Your options include:
1. Simply allowing this to be the bigger PR as it became, or
2. Moving some or all of the safe template work to a separate PR that either your team gets over the finish line, or that you transfer over to T&L so we don't lose all the work you've done thus far.

As you mentioned, T&L will take care of the resolving merge conflicts with #12772.

Thanks again for all the work.  Hopefully it helped familiarize you with the tools we are using for safe templates in general.

e-kolpakov commented:
@robrap Thank you for pointing out suboptimal solutions to template linting violations and for the review as a whole.

I'm fine with this PR being this size.

Yes, I've got some understanding on how to use HmtmUtils and friends - that was helpful.

Regarding failed build (you haven't mentioned that, but I believe it is implied): looks like moving plain HTML chunks in discussion/utils.js to templates have some disastrous effect on some non-discussion related tests - I'm trying to fix it right now.

robrap commented:
@e-kolpakov: I hadn't actually looked into the test failures.  That's too bad.  Does it mean we had suboptimal tests that were overly dependent on implementation?  I'm just not clear on how this would break non-discussion related tests.  

Good luck and let me know how and if I can help.  Thanks.

e-kolpakov commented:
@robrap In two words, it some tests use discussion module/XBlock in order for them to work, underscore fixtures must be added to a page. AFAIK, all discussions tests do that; I've added new templates both to actual production code and test helper, but for some reason some tests still don't get them.

The problem might have something to do with the fact that one of replaced templates is used in $.ajax.error handler; so any test that previously had discussions just broken now crash with JS error.

I'm looking into it right now, and it looks like there are two options: 
1. Make sure all tests that depend on discussion (even implicitly, like `TeamsTab` tests) have discussions template fixtures installed.
2. Just patch the error handler so that it don't crash in lieu of required template.

e-kolpakov commented:
@robrap Basically following approach number 1 was harder than expected, but approach number 2 worked right away, so I'm going to stick with it.

robrap commented:
Sounds fine. If you get a chance to remind me what approach 2 was, that would help. Thanks.

robrap commented:
Ignore. I just saw it was referring to your new comment.

e-kolpakov commented:
jenkins run python

e-kolpakov commented:
jenkins run a11y

e-kolpakov commented:
jecnkins run lettuce

e-kolpakov commented:
jenkins run lettuce

e-kolpakov commented:
jenkins run bokchoy

bjacobel commented:
👍  from me.

@e-kolpakov, thank you for all your work on this!!

e-kolpakov commented:
@bjacobel Thanks!

@robrap Friendly reminder to cast a +1 when you're satisfied with the code in this PR :)

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -444,19 +463,52 @@
+                // markdownWithHighlight requires unescaped text, so we must unescape it.
+                // safe-lint: disable=javascript-escape
+                elementContent = _.unescape(edx.HtmlUtils.HTML(elementContent));
+            }
```
I've got a couple of questions about this code: 

1.  Since `element.html()` is coming out of the DOM, it needed to be safe before whatever it was made it into the DOM, otherwise it is an unsafe way to pass this data around.  
2. What is the case where this is double-escaped (which is what I presume you must mean by `alreadyEscaped`, and could we simply avoid double-escaping?
3. Because we are using `element.html()`, characters like `&` are going to be escaped already (i.e. `&amp;), so wouldn't everything need to be unescaped if all of this is going to be escaped again?
4. I'm not sure why the `disable` comment above is needed, nor why the call to `HtmlUtils.HTML` is needed inside the code to `_.unescape()`.

I don't know what the appropriate code here should be, because I'm unclear about what is unique about the single call to convertMath where alreadyEscaped is set to true.  Let me know if you need more help working this out.

e-kolpakov replied:
@robrap This entire block of code turned out to be an attempt to compensate for the `converter.makeHtml` quirk of stripping html tags in post body, but escaping them in code blocks. So this caused both of them to be preserved and properly escaped, but in turn introduced more bugs in profile view, responses and comments.

So, I decided to stop fighting `converter.makeHtml` and just allow it to behave like it was; that means any html tags (including `script`) outside of code blocks will be stripped. If it is a UX regression - let's have it fixed as a separate PR/ticket/etc., as current master behaves exactly like that.

Answering the question (just to prove I'm sane and had a good idea why it should work like that :)):

1. It was safe - it came from `DiscussionThreadShowView.prototype.renderTemplate`
2. Html tags in code blocks was double-escaped (see explanation above)
3. That was a bug in this implementation.
4. disable - template linter reports all `_.escape` and `_.unescape`. Call to `HtmlUtils.HTML` - honestly - no idea; it shouldn't be there, but somehow it did what I wanted even with it (probably it just called toString on it).

robrap replied:
So you are saying that the following works:
```
elementContent = _.unescape(edx.HtmlUtils.HTML(elementContent));
```
But, the following does not work?:
```
elementContent = _.unescape(elementContent);
```
The call to HTML() should be a no-op, so I'm not sure why you would need it.  That's all I'm saying.

e-kolpakov replied:
@robrap No, I shouldn't have added `edx.HtmlUtils.HTML`, but it worked anyway and I haven't noticed it was there. And it worked exactly because call to `edx.HtmlUtils.HTML` was a no-op.

It wasn't needed, I just added it unintentionally (probably copied from some other place).

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -444,19 +463,52 @@
+            // 2. imagesToReplace.after(edx.HtmlUtils.interpolateHtml(
+            //     edx.HtmlUtils.HTML("<em>{text}</em>"), {text: gettext("image omitted")}
+            // ).toString())
+            // It does not report if it is written on single line, but than it becomes line too long jshint violation
```
I appreciate the explanation, but I think you could probably drop this comment.

e-kolpakov replied:
@robrap I believe the next guy to touch this code will try to optimize it and hit the issue I've faced, so I would rather keep the comment explaining why the code is a bit more complicated than it should be.

robrap replied:
@e-kolpakov: My opinion is that this is helpful, but verbose and not the best place for this.  We'd need the same verbose comment every place we use ensureHtml() to ensure we have an HTML Snippet.

I added the following JIRA ticket: https://openedx.atlassian.net/browse/TNL-4983

My recommendation would be to either not have anything in code (just in the Read the Docs), or to add a one-liner above 511 as follows:
```
// See TNL-4983 for an explanation of why the linter requires ensureHtml()
```

Let me know what you think of this proposal.

e-kolpakov replied:
@robrap makes sense

robrap commented on lms/templates/discussion/_underscore_templates.html:
```diff
@@ -1,17 +1,29 @@
+<%page expression_filter="h"/>
 <%namespace name='static' file='../static_content.html'/>
-<%! import json %>
+<%! from openedx.core.djangolib.js_utils import dump_js_escaped_json, js_escaped_string %>
```
`dump_js_escaped_json` is unused?

robrap commented:
@e-kolpakov: I get that this ballooned.  Your efforts are greatly appreciated.  Thank you!

e-kolpakov commented:
jenkins run js please

robrap commented on common/static/common/js/discussion/views/discussion_thread_show_view.js:
```diff
@@ -38,32 +38,34 @@
             DiscussionThreadShowView.prototype.convertMath = function() {
-                var element;
-                element = this.$(".post-body");
-                element.html(DiscussionUtil.postMathJaxProcessor(DiscussionUtil.markdownWithHighlight(element.text())));
```
@e-kolpakov: I still think that `alreadyEscaped` code is odd, and means something went a bit awry.  I think part of the problem is in this line that went away.  This used to call `element.text()`, and you now have it as 'element.html()' in convertMath().  I'm not sure if that is the only change needed, or if different calls were made in different contexts?

e-kolpakov replied:
@robrap yes, all the calls previously did `text()`, but combined with other changes it resulted in XSS, so I changed that to `html()`. The latest version have `text()` restored, at a cost of stripping HTML tags in post (thread/response/comment) body.

This `convertMath` gives a lot of trouble.

robrap replied:
@e-kolpakov: The latest makes more sense to me than the `alreadyEscaped` code?  Are you happy enough with the latest, or are you just hoping I'll be appeased and you can move on? ;)

e-kolpakov replied:
@robrap current version makes more sense than `alreadyEsscaped` (which also had a bug in - profile page was broken). Yes, I'm more happy with it, than previous; ideally it should also preserve tags in post body itself, but as I've already said looks like it is "by design" decision in markdown converter + it is existing behavior

robrap commented:
@e-kolpakov: Seems something went wrong with the tests.  Only 1 check was run.  You want to run them all?

robrap commented:
jenkins run all

robrap commented:
@jzoldak: All of the checks were running earlier, but now we are down to just quality.  I just ran all, and only quality came back.  Maybe jenkins just has a queue?  I'll ask on HipChat, but leaving this note for @e-kolpakov.

jzoldak commented:
Closing and re-opening this PR to work around an issue with the Jenkins GitHub Pull Request Builder cache.

robrap commented:
jenkins run quality

e-kolpakov commented:
jenkins run quality without errors :)

e-kolpakov commented:
jenkins run bokchoy without error :) (it worked last time :))

jzoldak commented:
@e-kolpakov if it's "test_prerequisite_course_save_successfully" that's killing you on the bok-choy side, rebase to get #12982

robrap commented on common/static/common/js/discussion/utils.js:
```diff
@@ -353,84 +356,107 @@
+            *       either body tags are not escaped, or code tags are double escaped, unless we specifically unescape
+            *       html tags in code templates and pass escaped code here.
+            *       Current implementation does not do that and just strips tags in body (looks like it is by design)
+            */
```
@e-kolpakov: Can we work through this comment here and you can update when you squash, assuming you don't want to run tests again and again?  There is a typo, and I think I would comment this differently.

What do you think of:
```
           /*
            * converter.makeHtml and HTML escaping:
            * - converter.makeHtml is not HtmlSnippet aware, so we must pass unescaped text raw text
            * - converter.makeHtml strips html tags in post body and escapes in code blocks by design.
            *    HTML tags are not supported.  Only markdown is supported.
            */
```

Is this accurate?  I'm not sure why there is any need for a TODO here.

robrap commented:
@e-kolpakov: Nice job.  Let's just work out an update to that comment for your squash commit.  Thanks for all the work.  👍

e-kolpakov commented:
jenkins run python

e-kolpakov commented:
jenkins run bokchoy without error please

bjacobel commented:
jenkins run bokchoy

e-kolpakov commented:
@bjacobel I believe this is a real test failure - something related to markdown rendering. I just haven't had time to fix it today.

e-kolpakov commented:
jenkins run python

e-kolpakov commented:
@bjacobel @robrap FYI I've fixed the problem and previous build succeeded: https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/23032/ - if you would like to make one last pass through this - it is the time.

If the CI build succeeds I'll merge it tomorrow morning my time (late night in Boston).

e-kolpakov commented:
jenkins run python without errors please

e-kolpakov commented:
@bjacobel @robrap @bradenmacdonald The build is green and we're ready to go, but merging something  Friday evening is a sure recipe for work weekend. Thus I believe it would be safer to merge it Monday morning, unless the release is cut today, or there are something else that requires merging it today.

Let me know if you have arguments for merging it today.

bradenmacdonald commented:
@bjacobel @robrap Thanks for your help getting this to land! :-)

robrap commented:
Of course.  Thanks to @e-kolpakov for all the safe template work as well.  This didn't make Friday's release cut, so it will go out next week.

e-kolpakov commented:
@robrap @bjacobel thank you for review!
